runners/direct-java/src/main/java/org/apache/beam/runners/direct/QuiescenceDriver.java

```diff
@@ -70,6 +71,8 @@ public static ExecutionDriver create(
   private final Map<AppliedPTransform<?, ?, ?>, ConcurrentLinkedQueue<CommittedBundle<?>>>
       pendingRootBundles;
   private final Queue<WorkUpdate> pendingWork = new ConcurrentLinkedQueue<>();
+  private final Map<AppliedPTransform<?, ?, ?>, Collection<CommittedBundle<?>>> inflightBundles =
+      new ConcurrentHashMap<>();
```

Comment:
We collect here bundles and AppliedPTransforms that have started to process bundle, but have not completed it yet. The reason for that is that the bundle processing might change output watermark of a PTransform before enqueuing the resulting bundle to pendingUpdates of downstream PTransform, which can lead to watermark being updated past the emitted elements.

Reply:
The watermark should be held by all timestamps in the bundle, until the bundle is completed and committed.

Can you move your GitHub review comment into a code comment that explains this field?

Reply:
Moved the comment to code. One more note - the issue arrises, because of how output watermark of PTransform is directly connected to input watermark of downstream PTransform. Maybe a more "technically correct" solution would be to attach output watermark updates to the bundle processing, so that bundle life-cycle would become:
 - start bundle
 - finish bundle (and enqueue and resulting bundles to pendingUpdates in downstream PTransform)
 - update output watermark
But that would most probably require a more complex refactor.

Reply:
Yes that sounds about right.